systems/controllers/constrained_lqr_controller.cc

```diff
@@ -1,103 +1,95 @@
-      this
-          ->DeclareVectorInputPort(OutputVector<double>(
-              num_positions_, num_velocities_, num_efforts_))
-          .get_index();
+  input_port_info_index_ = this->DeclareVectorInputPort(
+      OutputVector<double>(plant_.num_positions(),
+          plant_.num_velocities(), plant_.num_actuators())).get_index();
 
   // Output port that outputs the efforts
-  output_port_efforts_index_ =
-      this->DeclareVectorOutputPort(TimestampedVector<double>(num_efforts_),
-                                    &ConstrainedLQRController::CalcControl)
-          .get_index();
+  output_port_efforts_index_ = this->DeclareVectorOutputPort(
+      TimestampedVector<double>(plant_.num_actuators()),
+          &ConstrainedLQRController::CalcControl).get_index();
 
   // checking the validity of the dimensions of the parameters
-  DRAKE_DEMAND(q.size() == num_positions_);
-  DRAKE_DEMAND(u.size() == num_efforts_);
   DRAKE_DEMAND(lambda.size() == num_forces_);
-  DRAKE_DEMAND(Q.rows() == num_positions_ + num_velocities_);
+  DRAKE_DEMAND(Q.rows() == plant_.num_positions() + plant_.num_velocities());
   DRAKE_DEMAND(Q.rows() == Q.cols());
-  DRAKE_DEMAND(R.rows() == num_efforts_);
+  DRAKE_DEMAND(R.rows() == plant_.num_actuators());
   DRAKE_DEMAND(R.rows() == R.cols());
 
-  // Creating the full state vector (Velocities are zero as it is a fixed point)
-  VectorXd x(num_states_);
-  VectorXd v = VectorXd::Zero(num_velocities_);
-  x << q, v;
-  desired_state_ = x;
-
-  // Computing the full Jacobian (Position and contact)
-  MatrixXd J_tree_qdot, J_tree_v, J_contact_qdot, J_contact_v;
-  KinematicsCache<double> k_cache = tree_.doKinematics(q);
+  auto J_active_v = evaluators_.EvalActiveJacobian(context);
 
-  // Two separate Jacobians -- in terms of qdot and v are computed as the number
-  // of positions and velocities may be different (Quaternion base model)
-
-  // Position Jacobian
-  J_tree_qdot = tree_.positionConstraintsJacobian(k_cache, true);
-  J_tree_v = tree_.positionConstraintsJacobian(k_cache, false);
+  // convert to w.r.t. qdot one column at a time
+  MatrixX<AutoDiffXd> J_active_qdot(J_active_v.rows(), plant_.num_positions());
+  for (int i = 0; i < plant_.num_positions(); i++) {
+    AutoDiffVecXd v_i(plant_.num_velocities());
+    AutoDiffVecXd qdot = AutoDiffVecXd::Zero(plant_.num_positions());
+    qdot(i) = 1;
+    plant_.MapQDotToVelocity(context, qdot, &v_i);
+    J_active_qdot.col(i) = J_active_v * v_i;
+  }
 
-  // Contact Jacobian (If contact information is provided)
-  if (contact_info_.num_contacts > 0) {
-    J_contact_qdot = autoDiffToValueMatrix(
-        contact_toolkit_->CalcContactJacobian(initializeAutoDiff(x), true));
-    J_contact_v = autoDiffToValueMatrix(
-        contact_toolkit_->CalcContactJacobian(initializeAutoDiff(x), false));
+  // Add quaternion constraints to F
+  // Constraints come from the kinematic fact that ||q|| = 1, linearized is
+  // dot(q_0, delta_q) = 0
```

Comment:
I didn't get what the dot(,) function means by reading it. Maybe we can change it to 
`<q_0, delta_q> = 0`
or
`q_0^T * delta_q = 0`
?

Reply:
Done